src/stream/src/executor/watermark_filter.rs

```diff
@@ -0,0 +1,477 @@
+
+                        // Take the global max watermark when scaling happens.
+                        if previous_vnode_bitmap != vnode_bitmap {
+                            current_watermark =
+                                Self::get_global_max_watermark(&table, watermark_type.clone())
+                                    .await?;
+                        }
+                    }
+
+                    if barrier.checkpoint {
+                        // Persist the watermark when checkpoint arrives.
+                        let vnodes = table.get_vnodes();
+                        for vnode in vnodes.iter_pos() {
+                            let pk = Some(ScalarImpl::Int16(vnode as _));
+                            let row = Row::new(vec![pk, current_watermark.clone()]);
+                            table.insert(row);
+                        }
+                        table.commit(barrier.epoch).await?;
+                    } else {
+                        table.commit_no_data_expected(barrier.epoch);
+                    }
+
+                    yield Message::Barrier(barrier);
+                }
+            }
+        }
+    }
+
+    fn build_watermark_filter_expr(
+        watermark_type: DataType,
+        event_time_col_idx: usize,
+        watermark: Datum,
+    ) -> ExprResult<BoxedExpression> {
+        new_binary_expr(
+            Type::GreaterThanOrEqual,
+            DataType::Boolean,
+            InputRefExpression::new(watermark_type.clone(), event_time_col_idx).boxed(),
+            LiteralExpression::new(watermark_type, watermark).boxed(),
+        )
+    }
+
+    async fn get_global_max_watermark(
+        table: &StateTable<S>,
+        watermark_type: DataType,
+    ) -> StreamExecutorResult<Datum> {
+        let watermark_iter_futures = (0..VIRTUAL_NODE_COUNT).map(|vnode| {
+            async move {
+                let pk = Row::new(vec![Some(ScalarImpl::Int16(vnode as _))]);
+                let watermark_iter = table.iter_with_pk_prefix(&pk).await?.fuse();
+                pin_mut!(watermark_iter);
+                let watermark_row = watermark_iter.next().await;
+
+                // The remote storage should only contains at most 1 watermark for each vnode.
+                let next_row = watermark_iter.next().await;
+                if next_row.is_some() {
+                    bail!("Should not have more than on value for a given vnode")
+                }
+
+                match watermark_row {
+                    Some(row) => {
+                        let row = row?;
+                        if row.size() == 1 {
+                            Ok::<_, StreamExecutorError>(row[0].to_owned())
+                        } else {
+                            bail!("The watermark row should only contains 1 datum");
+                        }
+                    }
+                    _ => Ok(None),
+                }
+            }
+        });
+        let watermarks: Vec<_> = join_all(watermark_iter_futures)
```

Comment:
I think we can `try_join_all` here. If we get `None` for one vnode, we are sure that the value is not initialized and can directly use the `min()`.

Reply:
This will only happen when we first start the cluster and only gains a little performance improvement. Seems to be an premature optimization.